### PR TITLE
Closes #2550 Add support for future dates by setting publication status to "forthcoming"

### LINKED ISSUE #2550: Add forthcoming functionality to `az_bluebook` citation style.
## Motivation

According to the BlueBook style guide `forthcoming` is a feature that is supported in the Blue Book style, however our version, and  https://github.com/citation-style-language does not implement it

I have created an issue for this on https://github.com/citation-style-language/styles/issues/6576

### PR DESCRIPTION
**Pull request by trackleft**
## Description

Style changes: https://www.diffchecker.com/rhRIQGj0/

Resulting display:
image

## Related issues
Closes #2550 

## How to test
Set site default citation style to Bluebook.
Add new publication of any of the following types:
```
article-journal article-magazine article-newspaper book broadcast chapter interview manuscript map patent personal_communication song speech thesis webpage
```

Set the issue date to be some time in the future.

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [x] Update experimental module
- **Minor release changes**
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by trackleft** [comment]
We might also add a configuration option, that allows the site owner to mark future publications as forthcoming if their publish date is in the future.

**Comment by trackleft** [comment]
Add translation support

File: modules/custom/az_publication/az_publication.module [hunk 1]
```diff
@@ -214,6 +214,9 @@ function az_publication_node_view(array &$build, EntityInterface $entity, Entity
         // Handle publication date if available.
         if (!empty($entity->field_az_publication_date->value)) {
           $blob->issued = _az_publication_publication_date_object($entity);
+          if ($blob->issued->{'status'} === 'forthcoming') {
+            $blob->status = 'forthcoming';
```
Comment (trackleft) [inline modules/custom/az_publication/az_publication.module:218]:
```suggestion
            $blob->status = t('forthcoming');
```

**Comment by trackleft** [comment]
Blocked by https://github.com/az-digital/az_quickstart/pull/2942#pullrequestreview-1762936667

**Review by danahertzberg — DISMISSED** [review]

File: modules/custom/az_publication/config/install/az_publication.az_citation_style.az_bluebook.yml [hunk 2]
```diff
@@ -3,5 +3,5 @@ status: true
...
...sition=\"subsequent\">\r\n          <group delimiter=\", \">\r\n            <group delimiter=\" \">\r\n              <choose>\r\n                <if type=\"book\" match=\"any\">\r\n                  <text variable=\"volume\"/>\r\n                </if>\r\n              </choose>\r\n              <text macro=\"author-short\"/>\r\n            </group>\r\n            <group delimiter=\" \">\r\n              <text value=\"supra\" font-style=\"italic\"/>\r\n              <text value=\"note\"/>\r\n              <text variable=\"first-reference-note-number\"/>\r\n              <text macro=\"at_page\"/>\r\n            </group>\r\n          </group>\r\n        </else-if>\r\n        <else>\r\n          <group delimiter=\", \">\r\n            <group delimiter=\" \">\r\n              <choose>\r\n                <if type=\"book\" match=\"any\">\r\n                  <text variable=\"volume\"/>\r\n                </if>\r\n              </choose>\r\n              <text macro=\"author\"/>\r\n            </group>\r\n            <text macro=\"source\"/>\r\n            <text macro=\"access\"/>\r\n          </group>\r\n        </else>\r\n      </choose>\r\n    </layout>\r\n  </bibliography>\r\n</style>"
```
Comment (trackleft) [inline modules/custom/az_publication/config/install/az_publication.az_citation_style.az_bluebook.yml:6]: This adds `<date variable=\"issued\" prefix=\" \">\r\n`

**Comment by trackleft** [comment]
Interestingly we are [still overriding the bluebook style ](https://github.com/az-digital/az_quickstart/blob/issue/2550-bluebook-forthcoming/modules/custom/az_publication/config/install/az_publication.az_citation_style.az_bluebook.yml)in this repo in order to [add biblio support](https://github.com/citation-style-language/styles/pull/6634).

Additionally, the client I am implementing this for has additional overrides.
They remove the max authors listed settings for both citation, and bibliography elements to show all authors.

File: modules/custom/az_publication/config/install/az_publication.az_citation_style.az_bluebook.yml (hunk 2 above)
Comment (trackleft) [reply modules/custom/az_publication/config/install/az_publication.az_citation_style.az_bluebook.yml:6]: https://github.com/citation-style-language/styles/issues/6805

Comment (bberndt-uaz) [reply modules/custom/az_publication/config/install/az_publication.az_citation_style.az_bluebook.yml:6]: https://github.com/citation-style-language/styles/pull/6807

**Review by bberndt-uaz — APPROVED** [review]

**Comment by danahertzberg** [comment]
Troy will look into other citations to ensure styling is correct before approving

**Review by sanjanans — APPROVED** [review]

**Comment by tadean** [comment]
Citation style language contains an automatically localized `forthcoming` term: https://github.com/citation-style-language/locales/blob/master/locales-en-US.xml#L78 which isn't being used in the style customization.

It would be good if we can leverage this somehow, but checking the logic of the event date appears difficult. CSL 1.0 doesn't appear to handle the case well of a user-entered value needing to be a term.

**Comment by tadean** [comment]
APA, CMOS, and IEEE make use of the `status` variable also. AMA and MLA do not appear to currently.

Notably, APA also seems unsure whether the status should be printed as-is, or as a term: https://github.com/citation-style-language/styles/blob/a3d9a63426d2390068b4c98da6f48bd4ce73b257/apa.csl#L269 (presently it prints the raw term)

**Review by tadean — APPROVED** [review]

**Pull request merged**